iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp

```diff
@@ -133,11 +133,207 @@ LogicalResult clearStruct(OpBuilder builder, Value structValue,
   return success();
 }
 
-LogicalResult convertFuncOp(IREE::VM::FuncOp funcOp,
-                            VMAnalysisCache &vmAnalysisCache) {
+Optional<Value> findRefByOrdinal(OpBuilder builder, Location location,
+                                 mlir::FuncOp &parentFuncOp, size_t ordinal) {
+  auto ctx = builder.getContext();
+
+  // Search block arguments
+  int refArgCounter = 0;
+  for (BlockArgument arg : parentFuncOp.getArguments()) {
+    assert(!arg.getType().isa<IREE::VM::RefType>());
+
+    if (arg.getType() == emitc::OpaqueType::get(ctx, "iree_vm_ref_t*")) {
+      if (ordinal == refArgCounter++) {
+        return arg;
+      }
+    }
+  }
+
+  // Search local refs
+  for (auto constantOp : parentFuncOp.getOps<emitc::ConstantOp>()) {
+    Operation *op = constantOp.getOperation();
+    if (!op->hasAttr("ref_ordinal")) continue;
+    if (op->getAttr("ref_ordinal")
+            .cast<IntegerAttr>()
+            .getValue()
+            .getZExtValue() == ordinal) {
+      // Get address of constant
+      auto ptrOp = builder.create<emitc::ApplyOp>(
+          /*location=*/location,
+          /*type=*/emitc::OpaqueType::get(ctx, "iree_vm_ref_t*"),
+          /*applicableOperator=*/StringAttr::get(ctx, "&"),
+          /*operand=*/constantOp.getResult());
+
+      return ptrOp.getResult();
+    }
+  }
+
+  return None;
+}
+
+/// Find a local ref of type `iree_vm_ref_t*` matching the ordinal of the given
+/// `IREE::VM::Ref` value.
+Optional<Value> findRef(OpBuilder builder, Location location,
+                        mlir::FuncOp &parentFuncOp, Value refResult) {
+  auto ctx = builder.getContext();
+
+  assert(refResult.getType().isa<IREE::VM::RefType>());
+
+  size_t ordinal;
+  if (Operation *op = refResult.getDefiningOp()) {
+    // refValue is the result of an operation.
+    auto operandOrResultNumber =
+        [](Value value,
+           Operation *op) -> Optional<std::pair<unsigned, std::string>> {
+      for (OpOperand &operand : op->getOpOperands()) {
+        if (operand.get() == value) {
+          return std::make_pair(operand.getOperandNumber(),
+                                std::string("operand_ordinal"));
+        }
+      }
+      for (OpResult result : op->getResults()) {
+        if (result == value) {
+          return std::make_pair(result.getResultNumber(),
+                                std::string("result_ordinal"));
+        }
+      }
+      return None;
+    };
+
+    auto result = operandOrResultNumber(refResult, op);
+    unsigned index = std::get<0>(result.getValue());
+    std::string attrName = std::get<1>(result.getValue());
```

Comment:
when this is a string ref it won't matter, but this would also be taking a copy here on each access - `auto &` would prevent that, but better to just make it a simple value type instead